### ISSUE TITLE
DISC: AI policy

### ISSUE THREAD

**Issue by matthewturk** [yt-project/yt#5429]
We are overdue in providing guidelines for AI-based contributions to yt.  I don't think we should write our own AI policy, but rather adopt one from another community project.

A few examples that @chrishavlin has identified:

- [AstroPy](https://github.com/astropy/astropy-project/blob/main/policies/ai-policy.md)
- [SciPy](https://scipy.github.io/devdocs/dev/conduct/ai_policy.html)
- [sklearn](https://scikit-learn.org/stable/developers/contributing.html#automated-contributions-policy)

And, @neutrinoceros pointed out that in the AstroPy pull request they have identified others as well:

https://github.com/astropy/astropy-project/pull/530

**Comment by jzuhone** [comment]
I'll read these.

**Comment by chrishavlin** [comment]
Some quick thoughts I've been kicking around:

I generally think that adopting (and attributing!) one of the AI policies above is a good idea and simpler than trying to write our own from scratch. But the harder question is how to enforce adherence to whatever policy we have in place. With that in mind, here are some points I feel strongly about (and ways we might address them):

### disclosure 

I probably lean farther in the no-AI camp that other folks... but wherever we land, what I know for sure is that I want to know how AI is being used. So I'd like to update the pull request and issue templates to include checkboxes and fields to disclose use of AI. Things I'd want in a disclosure section:
   
```
- [ ] I promise that I am a human and I agree to not use AI in communication with other humans in comments, replies and/or discussions (except in use as a translation aid). 
- [ ] I have read the yt AI Policy (link to it)

Please describe the use of AI in this pull request (write "none" if AI was not used, otherwise follow the disclosure section of the AI Policy (link) ): 

```

### size of PRs 

The size of PRs need to be reasonable, but this is true for both AI and human-composed PRs. I haven't reviewed what we actually have in our contributor guidelines yet, but perhaps it'd help to add a requirement that more substantial PRs are required to have corresponding Issues with a clear statement that unsolicited/unexpected large PRs without prior discussion likely won't be reviewed (with the exception of adding a new frontend). 

### guidelines for responding to suspected AI policy infractions 

This most likely falls under the existing Code of Conduct, but it'd be helpful to have some concrete guidelines on what to do if there is a suspected violation of the AI policy. e.g., maybe one option for an action to take: if you see a PR that looks like an AI policy violation, tag a reviewer and move on (this would require us having a list of reviewers willing be tagged).

**Comment by matthewturk** [comment]
I think disclosure is absolutely essential, no matter where we end up -- even if we were to have a blanket ban (which does not feel completely right to me) we'd need an affirmative disclosure.

My thinking is that there's some allowance between "glorified tab complete" and "fully-agentic rewriting of the codebase."

It's also been brought up (I think by Chris?) that there's a place for different levels of acceptance, given the disparate levels of familiarity with the code base.  For instance, I would be more willing to accept an agentically-guided PR from someone I know is deeply familiar with the code base and who I know read it and I might be for someone who is new.  I don't know if this is something that needs to be codified but it's something to consider.

**Comment by jzuhone** [comment]
> It's also been brought up (I think by Chris?) that there's a place for different levels of acceptance, given the disparate levels of familiarity with the code base. For instance, I would be more willing to accept an agentically-guided PR from someone I know is deeply familiar with the code base and who I know read it and I might be for someone who is new. I don't know if this is something that needs to be codified but it's something to consider.

Full disclosure, as an example of such a thing, I already have plans to create a new data container-type object, and I plan to use an agent to help with the code design. I am fully confident that I have the ability to design, code, and test this functionality without such help, but I am also fully confident that I have almost no time to do so--sadly the increasing levels of managerial responsibilities I have to attend to these days means that I have less and less time for actual fun things like improving yt. :(

So the motivation is entirely utilitarian on my part--I recognize that we all have different levels of and reasons for comfortability with this sort of development pathway which may be different than mine. But I do think we can and should be able to agree on some 1) requirements that a human actually engage with the code itself and review and test it on their own and 2) that everyone disclose exactly how exactly they used AI agents in development, if at all.

**Comment by zingale** [comment]
I like Matt's view.  I've been playing with codex in some of our codes by asking it to look over the code and write a github-issue style .txt file for an bugs it finds.  In this sense, it is like a static analysis tool (which we also run a lot).  It found some useful things and some stuff that no one would ever hit.  Asking it to look for easy performance gains has been useful (it suggested to me that `std::sqrt(std::sqrt(x))` is faster than `std::pow(x, 0.25)` on GPUs, and it is, by a lot...

The AMReX folks added an AGENTS.md to guide this and have an MCU / context7 server setup.

I still fix things / write code myself, but having another tool to double check my work is useful.

**Comment by matthewturk** [comment]
> (it suggested to me that `std::sqrt(std::sqrt(x))` is faster than `std::pow(x, 0.25)` on GPUs, and it is, by a lot...

sidenote: this is the kind of thing that I find super interesting and now I want to learn more about.  I feel like I could *guess* a few possibilities why it's the case, but if I looked I'd find out they're all wrong?

**Comment by jzuhone** [comment]
> > (it suggested to me that `std::sqrt(std::sqrt(x))` is faster than `std::pow(x, 0.25)` on GPUs, and it is, by a lot...
> 
> sidenote: this is the kind of thing that I find super interesting and now I want to learn more about. I feel like I could _guess_ a few possibilities why it's the case, but if I looked I'd find out they're all wrong?

It's almost certainly because `sqrt` is using some special, highly optimized code since it's a frequently used function whereas `pow` is something that has to be generic enough to accept a ton of potential exponents and therefore can't make the same optimizations that `sqrt` does.

**Comment by cphyc** [comment]
Sorry for the long, and quite opiniated comment!

Here are my own personal two cents to add to the discussion. They mostly pertain to **AI & community building**.

I consider PRs to be one of our first entry points for community building, as is reflected by our ever-growing number of “team members”. To me, that points towards a collection of human beings that share a common goal (develop yt), not a collection of agents merely contributing code. In that sense, I am personally very much against having first-time contributors rely on AI to navigate our codebase, because this is replacing interactions with the already-existing community (through Slack, for example). I think I join @matt point of view: new contributors do not know the code yet, and submitting a PR requires gaining knowledge of the codebase _and_ of the community. AI just interferes in that process.

The situation is different for more mature contributors, since they know both the community and the codebase. Here, AI may help them reach the same goal (as was alluded to by @jzuhone), and only speeds up the process.

Personally, I have also found myself asking LLMs to do some tasks, one of them being the draft for a Dyablo frontend #5390. However, I don't take any pride in it, because the success of this endeavour relies _a lot_ on preexisting work (in the case above, @neutrinoceros's work on having a clean skeleton for new frontends is key) and the work of all of us to write good documentation. It is convenient to allow oneself to use AI and prevent others; if a decision that applies equally to all contributors needs to be made, I would lean in favour of banning AI contribution rather than keeping my self-appointed privilege.

[EDIT: added a second point]
The second point is about AI in general. _Who_ is pushing for it and _why_ it is pushing needs, IMO, to be included in the discussion. I've used AI in the past few months because I'm under pressure to deliver (project proposals, draft pieces of code, because I need to *quickly* demonstrate that something is working). This is not what I (and most of us) signed up for when I joined academia, nor is it what I signed up for when I started contributing to open-source projects. However, I end up using AI nonetheless because:
1. I am lazy, and AI exploits my tendency not to use my brain to do things I could have done in a similar amount of time. Essentially, it is a machine to make money out of thinking. Just like social media captures our attention, AI captures our thinking. I do not want to contribute to this.
2. I need to deliver, and AI is pushed with exactly this promise of "10x programmers" and other super-efficient unicorns. In the long run, we won't be able to compete with GPU-powered text-spitting machines. So we either accept giving away our work of love, or we refuse to put a coin in the machine (quite literally) and CO2 in the atmosphere. I would favour the second option, acknowledging 100% that I would need to actively prevent myself from using these tools.

**Comment by brittonsmith** [comment]
I think it would be fine to open the AI policy with a statement that we would generally prefer new contributors not us AI for the reasons @cphyc cites. Pragmatically, I think astropy's is good because it's brief and practical. There is a focus on expectations we are placing on the PR author. I would add a disclosure statement as the first point. Somewhere we need a statement about PR length, though it's not clear to me that it's here. Perhaps that goes in contributor guidelines, which we could link to.

**Comment by jzuhone** [comment]
This brings the question--how do we decide when a PR is "too long"?

**Comment by zingale** [comment]
Note that GCC is debating this now too, and potentially arguing that LLM code is not copyrightable:
https://gcc.gnu.org/pipermail/gcc/2026-April/247945.html

**Comment by brittonsmith** [comment]
> This brings the question--how do we decide when a PR is "too long"?

I'll just throw something out there. First time contributing and the PR is not a new frontend: 100 lines. Even that feels quite long. 20?

**Comment by matthewturk** [comment]
> I'll just throw something out there. First time contributing and the PR is not a new frontend: 100 lines. Even that feels quite long. 20?

I think some "rules of thumb" might work but I'm hesitant, since I think there is an argument to be made for stuff like "first time PR: make it so that all the enums are correct" or something similar.  Or, "every call to this one function  needs to be fixed" etc.

**Comment by brittonsmith** [comment]
I think these should all be rules of thumb that serve as guidance for reviewers and authors. Discretion should be used by reviewers in this sort of situation. There is probably wording that can communicate the difference between changing the syntax of 20 like function calls and adding a block of 20 new contiguous lines.

Discussion here seems to have stalled. Either we have some consensus or this is going cold. I think it's time to draft something up.

**Comment by neutrinoceros** [comment]
> Discussion here seems to have stalled

Sorry I was busy being angry at the state of the world. I'm going to read through this thread now.

@chrishavlin 
> I probably lean farther in the no-AI camp than other folks...

Try me.

> tag a reviewer and move on (this would require us having a list of reviewers willing be tagged).

I genuinely wonder who might sign up for this voluntarily. I know it's coming from good intentions, but I'm really not sure this is practical.

@matthewturk 
> even if we were to have a blanket ban (which does not feel completely right to me) we'd need an affirmative disclosure.

Yes on disclosure. I'll add that the *only* reason I will not advocate for a blanket ban is that I don't trust contributors to not lie about AI usage if we were to ban it completely.


@cphyc 
> I need to deliver, and AI is pushed with exactly this promise of "10x programmers"

That is seriously distressful to read. I get you're being "pragmatic" here, but I'm urging everyone to take a step back and refuse giving into exploitative business logics. We won't escape the spiral of vertically imposed pressure by giving in to ever higher productivity demands. In short: workers of the world, unite.

@zingale
> it to look over the code and write a github-issue style .txt file for an bugs it finds. In this sense, it is like a static analysis too

except that static analysis tools, traditionally:
- have deterministic behavior
- run locally
- impose bounded requirements on CPU usage (and more generally, on resources needed to run them)
- are not built on the underpaid, but essential work of the world's poorest nations (neo-colonialism ?)

My point is: my position doesn't have anything to do with how *useful* these tools may or may not be. I know this is a tangent to the overall discussion. I'm just really uncomfortable with all the negative externalities of LLMs being hand waved, or worse, not mentioned at all.

> LLM code is not copyrightable

LLM are notoriously being used for licence-laundering too. This has been a risk from the start and I haven't seen any sign of regulation of any kind that would prevent this.


@brittonsmith 
> 100 lines. Even that feels quite long. 20?

[Research](https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/) indicates that 400 is a physical threshold for humans. That's roughly the average scope window of a brain, if you like. Maybe it cannot be a hard ceiling because we don't want contributors to skip tests just to get under the limit, but we should just state upfront that the larger a PR, the longer it'll take to review, and 100 total LoC is already in the ballpark of "too long to review overnight".


Now for the original part of my reply:

I haven't seen any actual defensive mechanism proposed, and everyone up to now seems to assume all actors are well intended and that we should remain welcoming under any circumstance, which, to my despair, I don't think is sustainable. I've seen activity that might not be intended as harassment, but very much feel like it when you're on the receiving end (maintainers). To ask concrete questions, what do we do when a single account
- opens and closes multiple close-to-identical PRs 10 times an hour ?
- pings maintainers on 10 different open issues asking if it's still up for grabs in a minute ?
- creates a large PR and never responds to reviewers questions ?
- drops unprompted, verbatim "summaries" for known problems, together with some LLM-generated fix strategy that doesn't fix anything/makes no sense, or proposes a purely additive approach to a structural issue ?

My answer to all of the above would be: report the user and block them at the org level. If we were just to "label and move on", we'd inevitably end up diluting the signal to noise ratio on our bug tracker, and, perhaps more importantly, we'd be wasting already scarce maintainer time being nice to robots.

**Comment by chrishavlin** [comment]
In case anyone wants to check out other ai policies: https://github.com/melissawm/open-source-ai-contribution-policies

**Comment by brittonsmith** [comment]
Sorry for taking another two weeks to come back to this. Enforcement should be a key point to this policy. The actions listed by @neutrinoceros above to me would constitute a warning and a ban or potentially even an unwarned ban. However, I would want a final action on that to require concurrence by at least two people. It will benefit no one to have unilateral decision making in this space.

This policy needs to protect the community as a whole. It needs to make it clear to any contributors what will be expected of them, to make them understand the impact of their contribution on the other community members (specifically in terms of time commitment), and to clarify what they should expect in return. The moment we start talking about maintainers we are on the wrong foot. This project relies on the effort of volunteers whose buy-in depends on whether they can be made to feel like they belong. It does a disservice to frame a policy around those who are "in" and everyone else.

I propose that we try now to agree on the key sections of a policy and then to start iterating on each. For example, I can see the following:

- policy on AI use in contributions (how much if any, disclosure, requirement of understanding, etc)
- guidelines for PRs (length, scope, etc)
- enforcement

As I write these out, I'm not sure they all belong together in an AI policy or just in a broader contribution policy. Either way, I think we need to establish policy for them all.

**Comment by chrishavlin** [comment]
I agree with @brittonsmith : the PR guidelines belong in the general contribution policy. So just going to focus on AI specific policy here. 

After re-reading the full thread here and reading through astropy's ai-policy again... I actually think it covers our discussion well and we could use it with some minor modifications. I'll copy it here for reference:

<details>
<summary>astropy ai-policy (click for text)</summary>

# Usage of Generative AI ([link](https://github.com/astropy/astropy-project/blob/main/policies/ai-policy.md))

Astropy maintainers welcome contributions, including those developed with the assistance of generative AI tools. However, the use of such tools does not change our expectations for contributor responsibility, engagement, or code quality.
The following apply to all contributions.

1. **Human ownership and accountability:**
    1. Contributors are responsible for all submitted content, regardless of whether generative AI tools were used.
    2. Contributors must be able to explain all changes during review.
    3. **Licensing:** Contributors are responsible for ensuring all submitted code complies with Astropy's licensing, regardless of whether generative AI tools were used.
2. **Disclosure:** If generative AI tools were used for substantive portions of the contribution, contributors must describe their use in the change description. This also helps downstream citations comply with academic policies that may require disclosure of genAI usage.
3. **Authentic engagement:** The pull request process is collaborative and iterative. Contributors are expected to actively engage with reviewer feedback themselves. Copying and pasting replies to / from a generative AI tool does not count as engaging with the reviewer; contributors are expected to understand the changes they are proposing, not act as message courier for a third party.

> [!IMPORTANT]
> Autonomous workflows ("agents") are not human contributors. Astropy reviewers expect to interact with the human controlling the account, or we will close the PR.
4. **Consistency with Astropy conventions and existing style:** In the context of Generative AI, this particularly means conveying information (comments, documentation, etc) in a style the Astropy community uses, and that is appropriate and to the point.

> [!NOTE]
> This policy is in part inspired by the scikit-learn ["Automated Contributions Policy"](https://scikit-learn.org/stable/developers/contributing.html#automated-contributions-policy), which has additional details that may be useful in interpreting this policy.

</details>

I think that covers 90% of our discussion thread. What it is missing:

* modify disclosure statement: astropy's policy asks for disclosure of "substantive" use. Folks (including me) in our discussion thread want a bit more disclosure, so I think we should modify that sentence to require AI disclosure via a checkbox in the PR and issue templates noting an exception for language translation (see below). 
* no explicit note on enforcement. IMO we should add a `5. **Enforcement:**` section. See below. 
* expand the policy to explicitly cover both Issues and PRs (astropy's only mentions PRs)
* modify the opening paragraph: link to the existing contribution guidelines (which humans and robots must follow, we can update separately), rephrase to emphasize importance of new contributors not using AI .

Some thoughts on disclosure and enforcement:

### disclosure

For an AI disclosure checkbox, I think that if we want explicit disclosure we need to be a little less vague and emphasize levels of use. Here's an attempt at breaking it down into levels in the form of a PR template

```
By submitting this pull request, I agree to adhere to yt's AI policy (link), general Contribution Guidelines (link) and Code of Conduct (link).

I used AI for (leaving both boxes unchecked certifies that no AI was used in these changes):
[ ] low level assistance: code-completion, syntax lookup, explaining existing code.
[ ] substantive generation: substantive portions of code and/or documentation were written by AI then edited or submitted as is. This includes copying code from an AI chat or agentic coding. I have read and can explain every changed line.

I used the following AI tools: <write what you used>
``` 

Open to better ways to split out the level distinction here or modify the language and I'd want some slight modifications for an issue template, but this captures what I want to know. 

### enforcement 

In terms of what to put into an enforcement section, I think we should split this out into levels of violation based on the discussion above. 

* Low-level infractions: for example, a new contributor submits a PR that is probably AI and doesn't line up with disclosure checkbox. Response:  Ask for clarification on use of AI (fill out the checkbox if they didn't, ask for author to walk through the code changes), if no response in an amount of time (7 days?), close the PR with a note citing a violation of policies. 
* High-level infractions. Spam-like behavior (e.g., see [@neutrinoceros's comment](https://github.com/yt-project/yt/issues/5429#issuecomment-4351700612)). Response: immediately close issues and/or PRs. On confirmation from 2 yt project members, organization-level ban and github report. 


### drafting 

If it helps to speed up our discussion, I'm happy to draft an initial PR (but will wait a bit for responses).

**Comment by cphyc** [comment]
I think your approach is reasonable @chrishavlin. My main comment is Astropy's policy mostly pertains to contributions but doesn't describe what is expected from maintainers regarding AI. In particular, some of us may decide *not to* interact with AI-generated content, and I would suggest to explicitly mention it so that contributors know what to expect.

We've always been short on reviewers: allowing ourselves to favour human-generated content over AI-generated content in the context of this scarcity should probably be part of our guidelines.

**Comment by chrishavlin** [comment]
well I went ahead and wrote something up -- see #5541 . If it's too early for a PR, just LMK and i'll close it without argument. I would very much like to settle on something in the short term though, we are accruing more and more AI-based PRs, it would be nice to decide on an official strategy/policy. 

I'll give you all a chance to look first then can send around a note to slack and yt-dev (and maybe yt-users?) .
